### PR TITLE
ARTEMIS-6116 Fix tests for windows OS

### PR DESCRIPTION
**Pull request by GChuf**
Fixes tests on windows.

### CODE REVIEW

File: artemis-cli/src/test/java/org/apache/activemq/artemis/cli/commands/messages/ProducerThreadTest.java [hunk 1]
```diff
@@ -225,7 +225,7 @@ public void testBadMessagePropertyType() throws Exception {
       producer.setProperties(createJsonProperty("myType", "myKey", "myValue"));
       producer.applyProperties(mockMessage);
 
-      assertEquals("Unable to set property: myKey. Did not recognize type: myType. Supported types are: boolean, int, long, byte, short, float, double, string.\n", context.getStderr());
+      assertEquals("Unable to set property: myKey. Did not recognize type: myType. Supported types are: boolean, int, long, byte, short, float, double, string.\n", context.getStderr().replace("\r\n", "\n"));
```
Comment (jbertram) [inline artemis-cli/src/test/java/org/apache/activemq/artemis/cli/commands/messages/ProducerThreadTest.java]: This seems like a good place to use `System.lineSeparator()`.

File: pom.xml [hunk 2]
```diff
@@ -269,7 +269,6 @@
          -Djava.library.path="${activemq.basedir}/target/bin/lib/linux-x86_64:${activemq.basedir}/target/bin/lib/linux-i686"
          -Djgroups.bind_addr=localhost
          -Djava.net.preferIPv4Stack=true
-         -Dbasedir=${basedir}
```
Comment (jbertram) [inline pom.xml]: Can you elaborate on the problem this is causing? Several tests use `System.getProperty("basedir")` so I would expect removing this line would cause such tests problems.

Comment (GChuf) [reply pom.xml]: This one resolves the following warning:
`Warning:  The system property basedir is configured twice! The property appears in <argLine/> and any of <systemPropertyVariables/>, <systemProperties/> or user property.`
If I remember correctly, it was producing some kind of error on windows. Maybe I should have added more info in the commit message.

Comment (jbertram) [reply pom.xml]: FWIW, this has nothing to do with Windows as far as I can tell. The same warning message appears on Linux. I don't necessarily mind making this change, but it shouldn't be associated with ARTEMIS-6116 which is specifically about Windows.

**Comment by jbertram** [comment]
As far as I can tell, the two JDK8/IBM related commits aren't specific to Windows so they shouldn't be associated with ARTEMIS-6116 either.

File: artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java [hunk 3]
```diff
@@ -658,9 +666,13 @@ private void testSSLAutoReloadPemConfigSources(boolean useSymbolicLinks) throws
...
 
+      // Normalize paths to forward slashes
+      String sourceKeyPath = sourceKey.replace("\\", "/");
+      String sourceCertPath = sourceCert.replace("\\", "/");
+
       Files.write(serverPemConfigFile.toPath(), Arrays.asList(new String[]{
-         "source.key=" + sourceKey,
-         "source.cert=" + sourceCert
+         "source.key=" + sourceKeyPath,
+         "source.cert=" + sourceCertPath
```
Comment (jbertram) [inline artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java]: Are you sure this is strictly necessary? My testing indicates that it isn't.

File: pom.xml (hunk 2 above)
Comment (GChuf) [reply pom.xml]: I re-tested the build now on Windows without this change and I cannot see any errors related to this. Like I said, I thought this would solve an **error** on Windows but I misremembered - I was initially working on this more than a couple of months ago. I'm moving this to a separate issue.

**Comment by GChuf** [comment]
> As far as I can tell, the two JDK8/IBM related commits aren't specific to Windows so they shouldn't be associated with [ARTEMIS-6116](https://issues.apache.org/jira/browse/ARTEMIS-6116) either.

You're right - moving these to a separate issue as well.

File: artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java (hunk 3 above)
Comment (GChuf) [reply artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java]: Retested - In my case, reverting back to "sourceKey" and "sourceCert" without replacing "\\\\" results in the following error:
```
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.528 s -- in org.apache.activemq.cli.test.WebServerCLITest
[INFO] Running org.apache.activemq.cli.test.WebServerComponentTest
[ERROR] Tests run: 33, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 6.559 s <<< FAILURE! -- in org.apache.activemq.cli.test.WebServerComponentTest
[ERROR] org.apache.activemq.cli.test.WebServerComponentTest.testSSLAutoReloadPemConfigSources -- Time elapsed: 0.014 s <<< ERROR!
ootAppDataLocalTempjunit-6633549398662602348serverserver-key.pem:Users
        at org.apache.activemq.artemis.component.WebServerComponent.getStoreFile(WebServerComponent.java:377)
        at org.apache.activemq.artemis.component.WebServerComponent.addStoreResourceScannerTask(WebServerComponent.java:439)
        at org.apache.activemq.artemis.component.WebServerComponent.addStoreResourceScannerTask(WebServerComponent.java:459)
        at org.apache.activemq.artemis.component.WebServerComponent.createServerConnector(WebServerComponent.java:342)
        at org.apache.activemq.artemis.component.WebServerComponent.start(WebServerComponent.java:203)
        at org.apache.activemq.cli.test.WebServerComponentTest.startSimpleSecureServer(WebServerComponentTest.java:373)
        at org.apache.activemq.cli.test.WebServerComponentTest.testSSLAutoReloadPemConfigSources(WebServerComponentTest.java:677)
        at org.apache.activemq.cli.test.WebServerComponentTest.testSSLAutoReloadPemConfigSources(WebServerComponentTest.java:618)

[ERROR] org.apache.activemq.cli.test.WebServerComponentTest.testSSLAutoReloadPemConfigSourcesWithSymbolicLinks -- Time elapsed: 0.016 s <<< ERROR!
ootAppDataLocalTempjunit-10694756726840033654storestore-key.pemC:Users
        at org.apache.activemq.artemis.component.WebServerComponent.getStoreFile(WebServerComponent.java:377)
        at org.apache.activemq.artemis.component.WebServerComponent.addStoreResourceScannerTask(WebServerComponent.java:439)
        at org.apache.activemq.artemis.component.WebServerComponent.addStoreResourceScannerTask(WebServerComponent.java:459)
        at org.apache.activemq.artemis.component.WebServerComponent.createServerConnector(WebServerComponent.java:342)
        at org.apache.activemq.artemis.component.WebServerComponent.start(WebServerComponent.java:203)
        at org.apache.activemq.cli.test.WebServerComponentTest.startSimpleSecureServer(WebServerComponentTest.java:373)
        at org.apache.activemq.cli.test.WebServerComponentTest.testSSLAutoReloadPemConfigSources(WebServerComponentTest.java:677)
        at org.apache.activemq.cli.test.WebServerComponentTest.testSSLAutoReloadPemConfigSourcesWithSymbolicLinks(WebServerComponentTest.java:623)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
ootAppDataLocalTempjunit-6633549398662602348serverserver-key.pemes:618->testSSLAutoReloadPemConfigSources:677->startSimpleSecureServer:373 » IllegalArgument Store file does not exist: C:Users
ootAppDataLocalTempjunit-10694756726840033654storestore-key.pemcesWithSymbolicLinks:623->testSSLAutoReloadPemConfigSources:677->startSimpleSecureServer:373 » IllegalArgument Store file does not exist: C:Users
[INFO]
[ERROR] Tests run: 36, Failures: 0, Errors: 2, Skipped: 0
[INFO]
[ERROR]

```

File: artemis-cli/src/test/java/org/apache/activemq/artemis/cli/commands/messages/ProducerThreadTest.java (hunk 1 above)
Comment (GChuf) [reply artemis-cli/src/test/java/org/apache/activemq/artemis/cli/commands/messages/ProducerThreadTest.java]: Changed to `assertEquals("Unable to set property: myKey. Did not recognize type: myType. Supported types are: boolean, int, long, byte, short, float, double, string." + System.lineSeparator()`

File: artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java (hunk 3 above)
Comment (jbertram) [reply artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java]: OK. Could you do something like this instead:
```java
      Files.write(serverPemConfigFile.toPath(), Arrays.asList(new String[]{
         "source.key=" + sourceKey.replace("\\", "/"),
         "source.cert=" + sourceCert.replace("\\", "/")
      }));
```

Comment (GChuf) [reply artemis-web/src/test/java/org/apache/activemq/cli/test/WebServerComponentTest.java]: Of course, this looks cleaner. Changed.

**Comment by GChuf** [comment]
@jbertram you mentioned you tested my changes as well - what Windows version are you using? I am getting different results on Windows10 and Windows11 it seems.

**Comment by jbertram** [comment]
I was testing on Windows 10, but I went back and tested again after you responded and it looks like the slashes are indeed a real problem.

**Comment by GChuf** [comment]
@jbertram can you review this again and let me know if anything needs changing? I think I solved everything, tested on win10 and win11.
There's only 1 remaining test failing in Artemis Server (on my machines, anyway), but that's because running "main" without arguments hangs on windows instead of exiting:
`MainTest.testNull » Timeout testNull() timed out after 5 seconds`
Maybe we should solve that behavior in the server code, not just "fix" the test scenario.

Also, the last commit is solving an edge case. Error only happens when user does not have permissions to create symlinks.

p.s. does it make sense to create a github job to run builds/tests on Windows?

**Comment by GChuf** [comment]
Rebased on main.

**Comment by jbertram** [comment]
Most of this looks good. I only have qualms with two commits:

- "Release file handles in MessageSerializerTest" - this appears to be masking an actual leak, at least in regards to the `Producer`.
- "Replace ':' in filename with '_'" - this appears to have genuine upgrade problems. Lock files created with `:` will no longer be loaded once this code is changed. This change should include a note in `versions.adoc` as well as code to detect & rename existing files using `:`. Alternatively, it may be simpler to just detect the OS and adjust accordingly so that non-Windows use-cases didn't need to change.

I'll merge the rest but I'm dropping these two commits.
